lib/ex_webrtc/peer_connection/twcc_recorder.ex

```diff
@@ -0,0 +1,307 @@
+
+    defp round_to_250(val), do: div(val + 125, 250)
+  end
+
+  import Bitwise
+
+  alias ExRTCP.Packet.TransportFeedback.CC
+
+  @uint8_range 0..255
+  @int16_range -32_768..32_767
+
+  @max_seq_no 0xFFFF
+  @max_fb_pkt_count 0xFF
+  @max_ref_time 0xFFFFFF
+  @breakpoint 0x7FFF
+
+  @default_sender_ssrc 1
+
+  @type t() :: %__MODULE__{
+          media_ssrc: non_neg_integer(),
+          sender_ssrc: non_neg_integer() | nil,
+          base_seq_no: non_neg_integer() | nil,
+          start_seq_no: non_neg_integer() | nil,
+          end_seq_no: non_neg_integer() | nil,
+          timestamps: %{non_neg_integer() => float()},
+          fb_pkt_count: non_neg_integer()
+        }
+
+  # start, base - inclusive, end - exclusive
+  # start, end - actual range where map values might be set
+  # base - from where packets should be added to next feedback
+  # if end == start, no packets are available
+  defstruct media_ssrc: nil,
+            sender_ssrc: nil,
+            timer: Timer.new(),
+            base_seq_no: nil,
+            start_seq_no: nil,
+            end_seq_no: nil,
+            timestamps: %{},
+            fb_pkt_count: 0
+
+  @spec record_packet(t(), non_neg_integer()) :: t()
+  def record_packet(%{start_seq_no: nil, end_seq_no: nil} = recorder, seq_no) do
+    # initial case, should only occur once
+    timestamp = Timer.get_time(recorder.timer)
+
+    %__MODULE__{
+      recorder
+      | base_seq_no: seq_no,
+        start_seq_no: seq_no,
+        end_seq_no: seq_no + 1,
+        timestamps: %{seq_no => timestamp}
+    }
+  end
+
+  def record_packet(recorder, raw_seq_no) do
+    %__MODULE__{
+      base_seq_no: base_seq_no,
+      start_seq_no: start_seq_no,
+      end_seq_no: end_seq_no,
+      timestamps: timestamps
+    } = recorder
+
+    timestamp = Timer.get_time(recorder.timer)
+
+    # internally, we don't wrap the sequence number around 2^16
+    # so when receiving a new seq_num, we have to "unroll" it
+    seq_no = unroll(raw_seq_no, end_seq_no)
+    timestamps = Map.put_new(timestamps, seq_no, timestamp)
+    base_seq_no = if seq_no < base_seq_no, do: seq_no, else: base_seq_no
+
+    {start_seq_no, end_seq_no} =
+      cond do
+        seq_no < start_seq_no -> {seq_no, end_seq_no}
+        seq_no >= end_seq_no -> {start_seq_no, seq_no + 1}
+        true -> {start_seq_no, end_seq_no}
+      end
+
+    %__MODULE__{
+      recorder
+      | base_seq_no: base_seq_no,
+        start_seq_no: start_seq_no,
+        end_seq_no: end_seq_no,
+        timestamps: timestamps
+    }
+    |> remove_old_packets(timestamp)
+  end
+
+  defp unroll(seq_no, end_seq_no) do
+    end_rolled = end_seq_no &&& @max_seq_no
+    delta = seq_no - end_rolled
+
+    delta =
+      cond do
+        delta in -@breakpoint..@breakpoint -> delta
+        delta < -@breakpoint -> delta + @max_seq_no + 1
+        delta > @breakpoint -> delta - @max_seq_no - 1
+      end
+
+    end_seq_no + delta
+  end
```

Comment:
Uhh, that's a piece of a real math. TBH, I am not going to dive into this except one question, will current implementation correctly handle receiving sequence numbers in the following order: 1, 65534?

Reply:
Assuming that `seq_no == 65_534` and `end_seq_no == 1`, `delta` will be equal to
`65_533`, so it will fall into second arm of the `cond`, so the delta will become
`65_533 - 65_535 - 1 == -3`, which implies that the packet with `1` is older than packet with `65_534`, which is most likely the case, when packets like that are received consecutively. Btw, the test "packets wrapping around sequence number boundary" tests for something like this.